pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,225 @@
+	}
+	volumeClone, ok := clone.(*api.PersistentVolume)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+	}
+	return volumeClone, nil
+}
+
+func (ctx *persistentVolumeStateContext) updatePersistentVolumeStatus(phase api.PersistentVolumePhase) error {
+	if ctx.volume.Status.Phase == phase {
+		return nil
+	}
+
+	volumeClone, err := makeCloneForUpdate(ctx.volume)
+	if err != nil {
+		return err
+	}
+	volumeClone.Status.Phase = phase
+
+	updatedVolume, err := ctx.client.UpdatePersistentVolumeStatus(volumeClone)
+	if err != nil {
+		return fmt.Errorf("Unexpected error saving PersistentVolumeStatus: %+v", err)
+	}
+	ctx.volume = updatedVolume
+	ctx.volumeIndex.Update(ctx.volume)
+	return nil
+}
+
+func (ctx *persistentVolumeStateContext) getCurrentState() (persistentVolumeState, error) {
+	var state persistentVolumeState
+	for _, s := range ctx.states {
+		if s.isCurrentState(ctx) {
+			if state != nil {
+				return nil, fmt.Errorf("PersistentVolume[%s] is in more than one state: %v  %v", ctx.volume.Name, state, s)
+			}
+			state = s
+		}
+	}
+	if state == nil {
+		return nil, fmt.Errorf("PersistentVolume[%s] did not match any state: %v", ctx.volume.Name, ctx.volume)
+	}
+	return state, nil
+}
+
+// STATE OBJECTS
+//
+// See docs/design/persistent-volume-controller-pv-flow.png
+//     docs/design/persistent-volume-controller-pvc-flow.png
+
+// Pending volumes are either not bound or require provisioning.
+type persistentVolumePendingState struct{}
+
+// Bound volumes have a ClaimRef and the volume is fully provisioned
+type persistentVolumeBoundState struct{}
+
+// Released volumes have a ClaimRef, but the claim 404s in the API. User deleted the claim.
+// 4 possible outcomes from this phase:
+//    1.  Retain - PV is kept in Released phase and is not available for another claim
+//	  2.  Recycle - PV is scrubbed clean and the ClaimRef is removed (will match Pending above on sync)
+//	  3.  Delete - PV is removed from infrastructure and PV is deleted via API
+//	  4.  Failed - PV failed to Recycle or Delete.
+type persistentVolumeReleasedState struct{}
+
+// persistentVolumePendingState determines if a PersistentVolume should be Pending phase.
+// The PV is Pending when it is not bound (no ClaimRef) or it is bound (has ClaimRef) but is not fully provisioned.
+func (state *persistentVolumePendingState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	if ctx.volume.Spec.ClaimRef == nil {
+		return true
+	}
+	if isBeingProvisioned(ctx.volume) {
+		return true
+	}
+	return false
+}
+
+func (state *persistentVolumePendingState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumePending); err != nil {
+		return err
+	}
+
+	// launch Provisioner in separate thread / Goroutine
+	// lock on pv.name like the main control, guaranteeing 1 thread processing
+	// a PV at a time but allowing concurrent processing of multiple volumes
+	// TODO: make this work :)
```

Comment:
This will most probably involve update of the PV annotations. This should be done in the same update as updatePersistentVolumeStatus() above to save some traffic (and 'PV Updated' events).

I.e. this pseudocode looks nice, it won't be *that* nice in reality.